### PR TITLE
K8 upgrade guide

### PR DESCRIPTION
**Pull request by ChrisKujawa**
## Description

As requested I wrote the k8 upgrade guide. Feel free to add things, which are missing or ask questions when something is unclear. I think we can do this in a collaboration fashion to get this done early so feel free to push changes ;)

@afirth just want to check with you whether this guide makes sense for you.

## Related issues

closes https://github.com/zeebe-io/zeebe/issues/5582

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: docs/src/operations/upgrade.md
Comment (saig0) [inline docs/src/operations/upgrade.md:79]: Small typo.

```suggestion
If you using helm you can update the image tag in your values file.
```

File: docs/src/operations/upgrade.md
Comment (saig0) [inline docs/src/operations/upgrade.md:25]: Please correct all version numbers. `24.4` should be `0.24.4` :sweat_smile:

**Review by saig0 — APPROVED** [review]
Thanks Chris! :cake: 

The duplication of the section is not perfect but works for now. Hopefully, we can improve it in the new docs.

**Comment by Zelldon** [comment]
@npepinpe @afirth any thoughts? otherwise I will merge it

**Comment by npepinpe** [comment]
I'm reviewing it but cannot submit the review since I have no internet. Can we wait this morning and see if I manage to do it? Otherwise I can look into redoing it on my phone 😅

**Comment by Zelldon** [comment]
@npepinpe fine for me

File: docs/src/operations/upgrade.md
Comment (npepinpe) [reply docs/src/operations/upgrade.md:29]: May not be the exact right examples, but it might be good to add some (ingress being one of them), to give some ideas to users how you could interrupt incoming connections.

File: docs/src/operations/upgrade.md
Comment (npepinpe) [inline docs/src/operations/upgrade.md:60]: Having this line here makes it sound like only the querying needs to be done on the leaders, when we need to do everything on the leaders. Maybe put it higher, or start this paragraph with this? I would also add how to find the nodes that are leaders for which partition (e.g. topology), and that the query below can help them verify this information.

File: docs/src/operations/upgrade.md
Comment (npepinpe) [inline docs/src/operations/upgrade.md:127]: I think for standard dev users, it's not clear how they would rollback to a previous version and use the backup to restore from. But I don't know if we want to go in details here...maybe at least link to the relevant docs how to use a `VolumeSnapshot` (e.g. link directly to the relevant paragraph here https://kubernetes.io/docs/concepts/storage/volume-snapshots/#provisioning-volumes-from-snapshots). Is this something we can do with the helm chart though? I don't think you can specify the volumes or any info. correct me if im wrong @salaboy

**Review by npepinpe — COMMENTED** [review]
Thanks for this, looks good, just had minor suggestions

File: docs/src/operations/upgrade.md
Comment (Zelldon) [reply docs/src/operations/upgrade.md:29]: This was just a copy from the other section, should we add this also to the other section then?

File: docs/src/operations/upgrade.md
Comment (Zelldon) [reply docs/src/operations/upgrade.md:60]: The idea was that we just do it on all nodes, such that we don't need to find the leaders. And when you query the endpoint you see the leaders of the node.

File: docs/src/operations/upgrade.md
Comment (Zelldon) [reply docs/src/operations/upgrade.md:127]: I did in line 35 and 36. Actually I would like to avoid these details :sweat_smile: The guides from the official docs are far better probably. And if we want to have this kind of guide then we need to do it by ourself which we haven't done as far as I know.

File: docs/src/operations/upgrade.md
Comment (npepinpe) [reply docs/src/operations/upgrade.md:29]: I guess ingress is pretty k8s specific - for users not using k8s, I can't really tell them how they should do this :sweat_smile:

**Comment by salaboy** [comment]
@Zelldon this is quite awesome.. because I wasn't added as a reviewer I hadn't had the chance to look at it yet, but if you guys give me one more day I might be able to take a look at it in more details to see if I have any comments to add.

File: docs/src/operations/upgrade.md
Comment (npepinpe) [reply docs/src/operations/upgrade.md:60]: Maybe we can pull this up so users are aware of this (doing it on all pods) before we give the steps? So something like

```md
##### From broker version >= 0.24.4

The [Partitions Admin Endpoint](#partitions-admin-endpoint) can be used to query the status of the partitions and perform operations for the upgrade. First we need stop the workflow processing, trigger a snapshot creation, then query the broker status repeatedly to ensure all conditions are met.

To do so, you need to do the following **for all brokers in your cluster**. If you are using the Helm chart, you can find all broker pods by running the following command: `kubectl get pods -l app.kubernetes.io/component=broker`.

Forward traffic to the broker's metric port (default 9600). This will allow you to send local requests via `curl` to the broker's admin endpoint.

```shell script
$ kubectl port-forward PODNAME 9600
Forwarding from 127.0.0.1:9600 -> 9600
Forwarding from [::1]:9600 -> 9600
```

Send post request to the `prepareUpgrade` endpoint. This will pause stream processing and will trigger snapshot creation.

```shell script
$ curl -X POST localhost:9600/actuator/partitions/prepareUpgrade
{"1":{"role":"FOLLOWER","snapshotId":null,"processedPosition":null,"processedPositionInSnapshot":null,"streamProcessorPhase":null},"2":{"role":"FOLLOWER","snapshotId":null,"processedPosition":null,"processedPositionInSnapshot":null,"streamProcessorPhase":null},"3":{"role":"LEADER","snapshotId":null,"processedPosition":-1,"processedPositionInSnapshot":null,"streamProcessorPhase":"PAUSED"}}
```

Then, query the status of the broker's partitions repeatedly using the following command:
```shell script
$ curl -X GET localhost:9600/actuator/partitions
{"1":{"role":"FOLLOWER","snapshotId":null,"processedPosition":null,"processedPositionInSnapshot":null,"streamProcessorPhase":null},"2":{"role":"FOLLOWER","snapshotId":null,"processedPosition":null,"processedPositionInSnapshot":null,"streamProcessorPhase":null},"3":{"role":"LEADER","snapshotId":null,"processedPosition":-1,"processedPositionInSnapshot":null,"streamProcessorPhase":"PAUSED"}}
```

until the following conditions are met for all partitions:
  * `processedPositionInSnapshot` is equal to `processedPosition`
  * all followers (partitions which report `"role": "FOLLOWER"`) have the same `snapshotId` as their leader (partitions which report `"role": "LEADER"`)

As a reminder, for a partition which is `FOLLOWER` on a given broker, another broker will be the `LEADER` for that same partition. To match that both leaders and followers have the same snapshot, you may need to query different brokers to get all the information.
```

I have a feeling I didn't write it so well, but you get the idea - I think it's helpful for users if we explain each step we're doing, and that the FOLLOWER/LEADER stuff may be spread across different brokers.

File: docs/src/operations/upgrade.md
Comment (npepinpe) [reply docs/src/operations/upgrade.md:127]: Makes sense :slightly_smiling_face: I'll defer to @salaboy if users can figure this out from just the links in the docs.

**Comment by Zelldon** [comment]
Sounds good @salaboy

**Review by npepinpe — APPROVED** [review]

**Comment by npepinpe** [comment]
ping @salaboy - we'd like to announce 0.25.0 this week, and we'd need to release this before - if you could take a look at it before Wednesday afternoon that'd be great :slightly_smiling_face:

**Comment by salaboy** [comment]
@npepinpe all good from my side. 
I was trying to figure out, if there is a simple way to detach the service from the stateful set, instead of saying that we need to remove the ingress, as with a normal deployment, as soon as the service is not bound to the deployments, traffic will not flow. The fact that you remove an ingress, doesn't mean that traffic will stop, as there might be workers inside the cluster, not using the ingress which is going to still be able to reach the brokers and modify stuff. 

This feels that it can be automated, I wonder if the cloud team has some kind of script already doing this? @afirth do you have something that does all the steps mentioned in this doc automatically?

**Comment by salaboy** [comment]
@npepinpe on my side helm charts 0.25.x are coming slow.. not sure they are going to be ready this week.

**Comment by npepinpe** [comment]
Hey everyone, we'll put this on hold as we rework the upgrade procedure. I want to apologize for the poor planning here which caused extra load on all of you. Please reach out to me if you have questions or concerns.

**Comment by Zelldon** [comment]
So we can close this?

**Pull request closed without merging**
